os/fs/smartfs/smartfs_utils.c

```diff
@@ -1211,22 +1230,21 @@ int smartfs_find_availableentry(struct smartfs_mountpt_s *fs, uint16_t *parentdi
 				direntry->dsector = psector;
 				direntry->doffset = offset;
 				direntry->flags = entry->flags;
-				*parentdirsector = direntry->dsector;
+				direntry->last_dsector = direntry->dsector;
 				return OK;
 			}
 
 			/* Not an available entry, check the next one */
 			offset += entrysize;
 			entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
 		}
-		direntry->dsector = psector;
+		direntry->last_dsector = psector;
```

Comment:
dsector here, right?

Reply:
No this is last_dsector only
If code comes here, it means no entry found.

So createdirentry below will update dsector.

It is the same as using dsector here then
Last_dsector = dsector outside the loop

Reply:
yes I understand what you mean, but please see below
```
while (psector != SMARTFS_ERASEDSTATE_16BIT) {
	while ((offset + entrysize) < readwrite.count) {
		/* Check if this entry is available */
		if (IS_AVAILABLE_ENTRY(entry)) {
			direntry->doffset = offset;
			direntry->flags = entry->flags;
			return OK;
		}

		/* Not an available entry, check the next one */
		offset += entrysize;
		entry = (struct smartfs_entry_header_s *)&fs->fs_rwbuffer[offset];
	}

	direntry->dsector = psector;
	psector = nextsector;
}

/* We couldn't find entry, so create new directory entry */
direntry->last_dsector = direntry->dsector; // can do this in smartfs_createentry
ret = smartfs_createentry(fs, direntry);
if (ret != OK) {
	fdbg();
}

return ret;
```

Reply:
This is what I mentioned
Same as current implementation, if you want I can do this hut it is only extra few lines.

Reply:
In my opinion, dsector represents current direntry, last_dsector is used for sector which is chain to current sector(newly added).
So use dsector and set last_dsector if it couldn't find entry is correct.
Also please note that ret is always ENOENT in line 1245, if is not necessary.